codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -375,13 +381,51 @@ public void write(int allowedBytes) {
                         writePromise = ctx.newPromise();
                         writePromise.addListener(this);
                     }
+                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
+                        // Free the partials as they are written
+                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
+                        writePromise.addListener(new ChannelFutureListener() {
+                            @Override
+                            public void operationComplete(ChannelFuture future) throws Exception {
+                                toFree.discardReadComponents();
+                            }
+                        });
+                    }
                     frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
-                        size == bytesWritten && endOfStream, writePromise);
+                            size == bytesWritten && endOfStream, writePromise);
                 } while (size != bytesWritten && allowedBytes > bytesWritten);
             } finally {
                 size -= bytesWritten;
             }
         }
+
+        @Override
+        public boolean merge(Http2RemoteFlowController.FlowControlled next) {
+            // Disucssion point: If the buffer were two adjacent slices of the same underlying buffer it would
+            // be super nice not to have to use a Composite to joing then together.
+            if (!(next instanceof FlowControlledData)) {
+                return false;
+            }
+            FlowControlledData nextData = (FlowControlledData) next;
+            // Discussion point: Just add the padding instead?
```

Comment:
Given that we're merging into frames I'm inclined to just pick one of the two values. Adding them together seems like a poor choice when the length of the data ~= padding length. How about max?